File: scripts/generate-blog-index.js
const fs = require('fs');
const path = require('path');

const postsDir = path.join(__dirname, '..', 'posts');
const outputJson = path.join(__dirname, '..', 'posts-index.json');

function parseFrontMatter(content) {
  const match = content.match(/^---\r?\n([\s\S]+?)\r?\n---/);
  if (!match) return { metadata: {}, body: content };
  
  const yamlContent = match[1];
  const metadata = {};
  
  yamlContent.split(/\r?\n/).forEach(line => {
    const colonIndex = line.indexOf(':');
    if (colonIndex > 0) {
      const key = line.slice(0, colonIndex).trim();
      let value = line.slice(colonIndex + 1).trim();
      
      // Strip outer quotes from value if present
      if ((value.startsWith('"') && value.endsWith('"')) || (value.startsWith("'") && value.endsWith("'"))) {
        value = value.slice(1, -1);
      }
      metadata[key] = value;
    }
  });
  
  return metadata;
}

function generateIndex() {
  console.log('Generating blog index...');
  
  if (!fs.existsSync(postsDir)) {
    console.log(`Directory ${postsDir} does not exist. Creating it.`);
    fs.mkdirSync(postsDir);
    fs.writeFileSync(outputJson, '[]');
    return;
  }
  
  const files = fs.readdirSync(postsDir);
  const posts = [];
  
  files.forEach(file => {
    if (path.extname(file).toLowerCase() === '.md') {
      const filePath = path.join(postsDir, file);
      const content = fs.readFileSync(filePath, 'utf-8');
      const metadata = parseFrontMatter(content);
      
      if (!metadata.title) {
        console.warn(`Warning: File ${file} is missing title in front-matter. Skipping.`);
        return;
      }
      
      const postId = path.basename(file, '.md');
      posts.push({
        id: postId,
        title: metadata.title,
        date: metadata.date || new Date().toISOString().split('T')[0],
        category: metadata.category || '未分類',
        description: metadata.description || '',
        thumbnail: metadata.thumbnail || 'assets/hero-bg.png'
      });
    }
  });
  
  // Sort posts by date descending
  posts.sort((a, b) => new Date(b.date) - new Date(a.date));
  
  fs.writeFileSync(outputJson, JSON.stringify(posts, null, 2), 'utf-8');
  console.log(`Successfully generated index with ${posts.length} posts at ${outputJson}`);
}

generateIndex();

File: scripts/generate_blog_index.py
import os
import json
import re
from datetime import datetime

posts_dir = r"C:\Users\User\.gemini\antigravity\scratch\awake-okinawa-hp\posts"
output_json = r"C:\Users\User\.gemini\antigravity\scratch\awake-okinawa-hp\posts-index.json"

def parse_front_matter(content):
    match = re.match(r"^---\r?\n([\s\S]+?)\r?\n---", content)
    if not match:
        return {}, content
    
    yaml_content = match.group(1)
    body = content[match.end():]
    metadata = {}
    
    for line in yaml_content.splitlines():
        if ":" in line:
            parts = line.split(":", 1)
            key = parts[0].strip()
            value = parts[1].strip()
            # Remove quotes
            if (value.startswith('"') and value.endswith('"')) or (value.startswith("'") and value.endswith("'")):
                value = value[1:-1]
            metadata[key] = value
            
    return metadata, body

def generate_index():
    print("Generating blog index in Python...")
    if not os.path.exists(posts_dir):
        print(f"Directory {posts_dir} does not exist. Creating it.")
        os.makedirs(posts_dir)
        with open(output_json, "w", encoding="utf-8") as f:
            f.write("[]")
        return
        
    posts = []
    for filename in os.listdir(posts_dir):
        if filename.endswith(".md"):
            file_path = os.path.join(posts_dir, filename)
            with open(file_path, "r", encoding="utf-8") as f:
                content = f.read()
            
            metadata, _ = parse_front_matter(content)
            if "title" not in metadata:
                print(f"Warning: {filename} has no title. Skipping.")
                continue
                
            post_id = os.path.splitext(filename)[0]
            posts.append({
                "id": post_id,
                "title": metadata.get("title"),
                "date": metadata.get("date", datetime.today().strftime('%Y-%m-%d')),
                "category": metadata.get("category", "未分類"),
                "description": metadata.get("description", ""),
                "thumbnail": metadata.get("thumbnail", "assets/hero-bg.png")
            })
            
    # Sort posts by date descending
    posts.sort(key=lambda x: x["date"], reverse=True)
    
    with open(output_json, "w", encoding="utf-8") as f:
        json.dump(posts, f, ensure_ascii=False, indent=2)
        
    print(f"Successfully generated index with {len(posts)} posts at {output_json}")

if __name__ == "__main__":
    generate_index()
